CompilerSource/parsing/lexer.h

```diff
@@ -127,6 +127,17 @@ class Lexer {
   const ParseContext *context;
   std::deque<OpenMacro> open_macros;
   ErrorHandler *herr;
+
+  /// Store the stringified versions of macros in a set so that they can be safely referred to by `std::string_view`.
+  ///
+  /// As `CodeSnippet`s point inside the source itself, it is not possible to make one that points to the `std::string`
+  /// created on stringifying a macro as that string doesn't exist within the source and can only be referred to within
+  /// the function it is created in and therefore when trying to access this string outside, it can cause a segfault.
+  ///
+  /// Thus, this set "extends" the lifetime of the stringified forms of macros by promoting their lifetimes to the
+  /// lifetime of the lexer itself. The reason a `std::unordered_set` is used is to deduplicate macros which stringify
+  /// to the same string.
+  Macro::StringifiedSet stringified_macros;
```

Comment:
Good call! Now, another option (that I do in JDI) is to let tokens optionally own their own content string. In JDI, a token owns strings like these, and also takes ownership of its content on copy (notably, when copying to a token buffer).

Reply:
One issue that comes to mind is that for almost all tokens in the source, you don't need to have them own their source. Thus, by adding the `mutable std::string owned_str;` to the `content` struct in JDI (which is what I assume stores the copied string), you're distributing the cost of storing the copied string across _all_ tokens in the source which means you end up using more memory as almost every token is now storing 24-32 (usually) extra bytes of an uninitialized `std::string`.

Reply:
It's true. But memory is cheap, and so is the initialization. This way, every string we compose lives as long as the lexer, so we rely on the lexer being short-lived... but it still has to live as long as every token produced from it.